File: LogAndroid/TrustarApp/src/main/kotlin/jp/co/toukei/log/trustar/rest/post/PicPost.kt
package jp.co.toukei.log.trustar.rest.post

import android.util.Base64
import jp.co.toukei.log.lib.encodeToBase64String
import jp.co.toukei.log.lib.util.RequestBodyJson
import jp.co.toukei.log.trustar.db.user.entity.IncidentalHeader
import jp.co.toukei.log.trustar.db.image.ImageSending
import jp.co.toukei.log.trustar.user.ClientInfo
import org.json.JSONObject
import third.jsonObj

class PicPost(
        private val clientInfo: ClientInfo,
        private val imageSending: ImageSending
) : RequestBodyJson {

    @JvmField
    val picId = imageSending.picId

    override fun jsonBody(): JSONObject {
        return jsonObj {
            "companyCd" v imageSending.companyCd
            "userId" v imageSending.userId
            "clientInfo" v clientInfo.jsonBody()

            "picId" v picId
            "picByteBase64String" v imageSending.picRaw.encodeToBase64String(Base64.NO_WRAP)

            val content = imageSending.content
            //this sucks.
            "picCls" v content.optString("picCls")
            "note1" v content.optString("note1")
            "note2" v content.optString("note2")
            "note3" v content.optString("note3")
            "note4" v content.optString("note4")
            "note5" v content.optString("note5")
        }
    }

    sealed class Content(
            @JvmField
            val picCls: String,
            @JvmField
            val note1: String?,
            @JvmField
            val note2: String?,
            @JvmField
            val note3: String?,
            @JvmField
            val note4: String?,
            @JvmField
            val note5: String?
    ) {

        fun toJson(): JSONObject {
            return jsonObj {
                "picCls" v picCls
                "note1" v note1
                "note2" v note2
                "note3" v note3
                "note4" v note4
                "note5" v note5
            }
        }

        class IncidentalSign(header: IncidentalHeader) : Content(
                "SIGN",
                header.allocationNo,
                header.allocationRowNo.toString(),
                header.uuid,
                "Incidental",
                null
        )
    }
}

File: LogiOS/TrustarApp/Services/Request/PicPost.swift
//
//  PicPost.swift
//  TrustarApp
//
//  Created by DionSoftware on 2022/03/05.
//  Copyright © 2022 DionSoftware. All rights reserved.
//

import Foundation

class PicPost: Codable {
    private var clientInfo: ClientInfo
    private var imageSending: ImageSending
    private var picId = Resources.strEmpty
    
    init(clientInfo: ClientInfo, imageSending: ImageSending) {
        self.clientInfo = clientInfo
        self.imageSending = imageSending
        self.picId = imageSending.picId
    }
    
    func result() -> Result {
        let rs = Result()
        rs.companyCd = imageSending.companyCd
        rs.userId = imageSending.userId
        rs.clientInfo = clientInfo
        
        rs.picId = picId
        rs.picByteBase64String = imageSending.picRaw.base64EncodedString()
        
        if let parsedData = try? JSONSerialization.jsonObject(with: imageSending.content.data(using: .utf8)!, options: []) as? [String:Any] {
            rs.picCls = parsedData["picCls"] as? String ?? ""
            rs.note1 = parsedData["note1"] as? String
            rs.note2 = parsedData["note2"] as? String
            rs.note3 = parsedData["note3"] as? String
            rs.note4 = parsedData["note4"] as? String
            rs.note5 = parsedData["note5"] as? String
        }
        
        return rs
    }
    
    class Result: Codable {
        var companyCd: String = ""
        var userId: String = ""
        var clientInfo: ClientInfo?
        
        var picId: String = ""
        var picByteBase64String: String = ""
        var picCls: String = ""
        var note1: String?
        var note2: String?
        var note3: String?
        var note4: String?
        var note5: String?
    }
    
    class Content {
        var picCls: String
        var note1: String?
        var note2: String?
        var note3: String?
        var note4: String?
        var note5: String?
        
        init(picCls: String,
             note1: String?,
             note2: String?,
             note3: String?,
             note4: String?,
             note5: String?) {
            self.picCls = picCls
            self.note1 = note1
            self.note2 = note2
            self.note3 = note3
            self.note4 = note4
            self.note5 = note5
        }
        
        func toJsonString() -> String {
            let jsonObj: [String: Any] = [
                "picCls": picCls,
                "note1": note1 ?? "",
                "note2": note2 ?? "",
                "note3": note3 ?? "",
                "note4": note4 ?? "",
                "note5": note5 ?? ""
            ]
            guard let theJSONData = try? JSONSerialization.data(withJSONObject: jsonObj,
                                                                options: .prettyPrinted) else {
                return ""
            }
            
            return String(data: theJSONData, encoding: .utf8) ?? ""
            
        }
        
        class IncidentalSign : Content {
            init (header: IncidentalHeader) {
                super.init (picCls: "SIGN",
                            note1: header.allocationNo,
                            note2: String(header.allocationRowNo),
                            note3: header.uuid,
                            note4: "Incidental",
                            note5: nil)
            }
        }
    }
}
